common/sampling.h

```diff
@@ -0,0 +1,110 @@
+    std::string cfg_negative_prompt;   // string to help guidance
+    float       cfg_scale     = 1.f;   // How strong is guidance
+
+    std::unordered_map<llama_token, float> logit_bias; // logit bias for specific tokens
+
+} llama_sampling_params;
+
+// per-sequence sampler state
+typedef struct llama_sampler_sequence_state {
+    float mirostat_mu; // mirostat sampler state
+    llama_grammar * grammar;
+} llama_sampler_sequence_state;
+
+// general sampler state
+typedef struct llama_sampling_state {
+    ~llama_sampling_state();
+
+    // parameters that will be used for sampling and when creating
+    // new llama_sampler_sequence_state instances
+    llama_sampling_params params;
+
+    // map of sequence ids to sampler states
+    std::unordered_map<llama_seq_id, llama_sampler_sequence_state> sequence_states;
+
+    // when non-NULL, new instances of llama_sampler_sequence_state
+    // will get a copy of the grammar here
+    // note: only the pointer is stored here, it is not a copy of
+    //       the grammar and shouldn't be freed
+    llama_grammar * grammar;
+} llama_sampling_state;
+
+#include "common.h"
+
+// Create a new sampling state instance.
+llama_sampling_state llama_sampling_state_init(
+        const struct gpt_params & params,
+                  llama_grammar * grammar = NULL);
+
+// Fetches the sampler state for the specified sequence id (defaults to 0).
+// If the state for that sequence id doesn't already exist, it will be created with
+// default values based on the parameters in the state argument.
+llama_sampler_sequence_state & llama_sampling_get_sequence_state(
+              llama_sampling_state & state,
+        const llama_seq_id           seq = 0);
+
+// Reset the sampler states for the supplied sequence id (defaults to 0).
+// This is necessary to reuse a sequence id or free memory used by sequences
+// that are no longer required.
+bool llama_sampling_state_reset(
+              llama_sampling_state & state,
+        const llama_seq_id           seq = 0);
+
+// this is a common sampling function used across the examples for convenience
+// it can serve as a starting point for implementing your own sampling function
+// Note: When using multiple sequences, it is the caller's responsibility to delete
+//       the item in params.sampler_state when a sequence ends and samplers that rely on
+//       state are being used.
+//
+// required:
+//  - ctx:    context to use for sampling
+//  - params: sampling parameters
+//
+// optional:
+//  - ctx_guidance:  context to use for classifier-free guidance, ignore if NULL
+//  - grammar:       grammar to use for sampling, ignore if NULL
+//  - last_tokens:   needed for repetition penalty, ignore if empty
+//  - idx:           sample from llama_get_logits_ith(ctx, idx)
+//  - seq:           sequence id to associate sampler state with
+//
+// returns:
+//  - token:      sampled token
+//  - candidates: vector of candidate tokens
+//
+llama_token llama_sample_token(
```

Comment:
```suggestion
llama_token llama_sampling_sample(
```

Reply:
I really like this for reasons other than just consistency. I've found it confusing in the past that `llama_sample_token` was an overload in `common` while there was also a `llama_sample_token` in `lamma.cpp`.